virtool/http/auth.py

```diff
@@ -102,11 +110,77 @@ async def authenticate_with_api_key(req: Request, handler: Callable, user_id: st
     return await handler(req)
 
 
+async def authenticate_with_b2c(req: Request, handler: Callable) -> Response:
+    """
+    Authenticate requests when req.app["config"].use_b2c is True.
+
+    If no id_token cookie is attached to request, redirect to /api/acquire_tokens
+
+    If id_token cookie is found, attempt to validate to gather user information from claims. If token is expired,
+    redirect to /api/refresh_tokens. If token is invalid for some other reason, redirect to /api/delete_tokens
+
+    find or create user based on token claims, then populate req["cient"] with user information and return the
+    response from the handler.
+
+    :param req: the request to handle
+    :param handler: the handler to call with the request if it is authenticated successfully
+
+    :return: the response
+    """
+    if req.cookies.get("id_token") is None:
+        auth_code_flow = await req.app["run_in_thread"](
+            req.app["b2c"].msal.initiate_auth_code_flow,
+            ["email"],
+            f"http://localhost:9950/api/oidc/acquire_tokens",
+            random_alphanumeric(8)
+        )
+        req.app["auth_code_flow"] = auth_code_flow
+        return HTTPFound(auth_code_flow["auth_uri"])
+
+    try:
+        token_claims = await validate_token(req.app, req.cookies.get("id_token"))
+    except ExpiredSignatureError:
+        return HTTPFound("/api/oidc/refresh_tokens")
+    except (JWTClaimsError, JWTError):
+        await update_jwk_args(req.app, req.cookies.get("id_token"))
+        return HTTPFound("/api/oidc/delete_tokens")
+
+    user_attributes = B2CUserAttributes(
+        display_name=token_claims["name"],
+        given_name=token_claims["given_name"],
+        family_name=token_claims["family_name"],
+        oid=token_claims["oid"]
+    )
+
+    user_document = await find_or_create_b2c_user(req.app["db"], user_attributes)
+
+    req["client"] = UserClient(
+        db=req.app["db"],
+        administrator=user_document["administrator"],
+        force_reset=False,
+        groups=user_document["groups"],
+        permissions=user_document["permissions"],
+        user_id=user_document["_id"],
+        authenticated=True
+    )
+
+    return await handler(req)
+
+
 @web.middleware
-async def middleware(req, handler):
+async def middleware(req, handler) -> Response:
+    """
+    Handle requests based on client type and authentication status.
+
+    :param req: the request to handle
+    :param handler: the handler to call with the request if it is authenticated successfully
+    :return: the response
+    """
     db = req.app["db"]
 
-    if req.path == "/api/account/login" or req.path == "/api/account/logout":
+    if (req.method, os.path.split(req.path)[0]) in PUBLIC_ROUTES \
```

Comment:
What is this doing?
```
os.path.split(req.path)[0]
```

Reply:
I wrote it this way to mimic the check for `PUBLIC_ROUTES` in `virtool.jobs.auth` this way it just checks that the first piece of the `req.path` is `/api/oidc`

Reply:
There is too much potential here for unintentionally making routes public that should be protected. Stick to a static list of method-route combos.